### PR TITLE
Implement a testMode for running wrangler's dev function via API 

### LINKED ISSUE #1419: Chore: refactor local-mode-tests to use unstable_dev
- We should prove `unstable_dev()`'s ability for integration testing by rewriting our local-mode tests to use it instead

### LINKED ISSUE #1432: 🐛 BUG: API/unstable_dev race condition when running several `unstable_dev` commands at the same time
### What version of `Wrangler` are you using?

wrangler@beta

### What operating system are you using?

Mac

### Describe the Bug

In CI, running tests with unstable_dev fails sometimes. Investigating why lead to the following discoveries:

This blows up with `TypeError: fetch failed`:
```
    workers.push(
            await wrangler.unstable_dev("src/basicModule.ts"),
            await wrangler.unstable_dev("src/basicModule.ts")
        );
```

This works:
```
        workers.push(
            await wrangler.unstable_dev("src/basicModule.ts", { port: 9001 }),
            await wrangler.unstable_dev("src/basicModule.ts", { port: 9002 })
        );
```

This fails with a jest timeout: 

```
        workers = await Promise.all([
            wrangler.unstable_dev("src/basicModule.ts", { port: 9001 }) as Worker,
            wrangler.unstable_dev("src/basicModule.ts", { port: 9002 }) as Worker,
        ]);
```

### PR DESCRIPTION
**Pull request by rozenmd**
Prior to this change, `wrangler.unstable_dev()` would only support running one instance of wrangler at a time, as Ink only lets you render one instance of React. This resulted in test failures in CI.

This change creates pure JS/TS versions of these React hooks:

- useEsbuild
- useLocalWorker
- useCustomBuild
- useTmpDir

As a side-effect of removing React, tests should run faster in CI.

What we're not supporting (in this PR, anyway):
- multi-worker support (you should be able to test multiple workers at the same time, but they won't know about each other)
- watching/rebuilds - this API is designed to be used only by tests

Follow-up task:
- [ ] [Make it possible to specify a path when fetching](https://github.com/cloudflare/wrangler2/issues/1752)

Closes #1432
Closes #1419

### CODE REVIEW

File: packages/wrangler/src/dev.tsx [hunk 1]
```diff
@@ -533,6 +534,251 @@ export async function startDev(args: StartDevOptions) {
...
+			});
+		}
+
+		const entry = await getEntry(
+			{ assets: args.assets, script: args.script },
+			config,
+			"dev"
+		);
+
+		if (config.services && config.services.length > 0) {
```
Comment (rozenmd) [inline packages/wrangler/src/dev.tsx:579]: I feel like all of these logger.warn/errors should be common, right @threepointone ?

File: packages/wrangler/src/pages/dev.tsx [hunk 2]
```diff
@@ -260,23 +260,17 @@ export const Handler = async ({
...
+	process.on("SIGINT", async () => {
 		CLEANUP();
-		stop();
+		await stop();
 	});
-	process.on("SIGTERM", () => {
+	process.on("SIGTERM", async () => {
 		CLEANUP();
-		stop();
+		await stop();
```
Comment (JacobMGEvans) [inline packages/wrangler/src/pages/dev.tsx:273]: Is this the answer to `waitUntil` from before lol?

File: packages/wrangler/src/__tests__/api-dev.test.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { unstable_dev } from "../api";
+
+jest.unmock("undici");
+
+describe("unstable_dev", () => {
+	it("should return Hello World", async () => {
```
Comment (rozenmd) [inline packages/wrangler/src/__tests__/api-dev.test.ts:6]: This test will fail until we find a solution for https://github.com/cloudflare/wrangler2/issues/1755

**Review by threepointone — APPROVED** [review]
tentatively approving. that's for pair reviewing this with me!

File: packages/wrangler/src/__tests__/api-dev.test.ts (hunk 3 above)
Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/api-dev.test.ts:6]: See https://github.com/cloudflare/wrangler2/pull/1762

**Review by petebacondarwin — APPROVED** [review]

File: fixtures/local-mode-tests/tests/dotenv.test.ts [hunk 4]
```diff
@@ -1,15 +1,19 @@
 import { spawnWranglerDev } from "./helpers";
+console.log(spawnWranglerDev);
```
[the remark this answers is not in the document]
Comment (rozenmd) [reply fixtures/local-mode-tests/tests/dotenv.test.ts]: Nah i've removed this test and added a follow-up to reimplement it with unstable_dev: https://github.com/cloudflare/wrangler2/issues/1756

File: fixtures/local-mode-tests/tests/headers.test.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
+import { unstable_dev } from "wrangler";
+
+describe("worker", () => {
+	let worker: {
+		fetch: (init?: RequestInit) => Promise<Response | undefined>;
+		stop: () => Promise<void>;
+	};
```
[the remark this answers is not in the document]
Comment (rozenmd) [reply fixtures/local-mode-tests/tests/headers.test.ts:7]: We do export:
```ts

/**
 *  unstable_dev starts a wrangler dev server, and returns a promise that resolves with utility functions to interact with it.
 *  @param {string} script
 *  @param {DevOptions} options
 */
export declare function unstable_dev(script: string, options?: DevOptions, apiOptions?: DevApiOptions): Promise<{
    stop: () => Promise<void>;
    fetch: (init?: RequestInit | undefined) => Promise<Response | undefined>;
    waitUntilExit: () => Promise<void>;
}>;
```

I just couldn't figure out how to make the types work within the mono repo 😕

**Pull request merged**